SYCL/ESIMD/slm_alloc.cpp

```diff
@@ -0,0 +1,181 @@
+// REQUIRES: gpu
+// UNSUPPORTED: cuda || hip
+//
+// RUN: %clangxx -fsycl %s -o %t.1.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.1.out
+//
+// Vary the test case by forcing inlining of the functions with slm_allocator:
+// RUN: %clangxx -fsycl -DFORCE_INLINE %s -o %t.2.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.2.out
+
+// This is end-to-end test for the slm_allocator API used together with the
+// slm_init. The call graph is:
+//         Test1(kernel) - uses slm_init(SLM_IN_KERNEL)
+//         /    \
+//        /      v
+//       /       bar - uses slm_allocator(SLM_IN_BAR)
+//      v
+//     foo - uses slm_allocator(SLM_IN_FOO)
+// Test1 kernel SLM usage is SLM_IN_KERNEL + max(SLM_IN_BAR, SLM_IN_FOO).
+// SLM offset returned by the slm_allocator in foo and bar is the same and is
+// SLM_IN_KERNEL bytes.
+// Bar uses slightly bigger SLM frame than foo. It modifies values (adds 10) in
+// SLM resulting from foo, plus appends coule more '100's.
+
+#include <iostream>
+#include <sycl/ext/intel/esimd.hpp>
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+using namespace sycl::ext::intel::esimd;
+using namespace sycl::ext::intel::experimental::esimd;
+
+using T = int;
+constexpr int LOCAL_SIZE = 4;
+constexpr int GLOBAL_SIZE = 8;
+
+constexpr int NUM_WGS = GLOBAL_SIZE / LOCAL_SIZE;
+
+constexpr int ELEM_SIZE = sizeof(T);
+
+constexpr int SLM_IN_KERNEL = LOCAL_SIZE * ELEM_SIZE;
+constexpr int SLM_IN_FOO = LOCAL_SIZE * ELEM_SIZE;
+constexpr int BAR_EXTRA_ELEMS = 2;
+constexpr int SLM_IN_BAR = SLM_IN_FOO + BAR_EXTRA_ELEMS * ELEM_SIZE;
+constexpr int SLM_TOTAL = SLM_IN_KERNEL + std::max(SLM_IN_FOO, SLM_IN_BAR);
+constexpr int BAR_MARKER1 = 10;
+constexpr int BAR_MARKER2 = 100;
+
+#ifdef FORCE_INLINE
+inline
```

Comment:
It is supposed to be FORCE_INLINE, but 'inline' alone may be treated as a recommendation to FE.
Should it be using 'always_inline' attribute, similar to this one: https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/esimd/detail/defines_elementary.hpp#L48